File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsPlayer.java
package nl.t64.game.rpg.screens.world.entity;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.glutils.ShapeRenderer;
import com.badlogic.gdx.math.Rectangle;
import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.constants.Constant;
import nl.t64.game.rpg.screens.world.entity.events.*;

import java.util.List;


public class PhysicsPlayer extends PhysicsComponent {

    private boolean isActionPressed;

    public PhysicsPlayer() {
        this.isActionPressed = false;
        this.boundingBoxWidthPercentage = 0.80f;
        this.boundingBoxHeightPercentage = 0.30f;
    }

    @Override
    public void receive(Event event) {
        if (event instanceof LoadEntityEvent loadEvent) {
            direction = loadEvent.direction;
            currentPosition = loadEvent.position;
            setBoundingBox();
        }
        if (event instanceof StateEvent stateEvent) {
            state = stateEvent.state();
        }
        if (event instanceof DirectionEvent directionEvent) {
            direction = directionEvent.direction();
        }
        if (event instanceof SpeedEvent speedEvent) {
            velocity = speedEvent.moveSpeed();
        }
        if (event instanceof ActionEvent) {
            isActionPressed = true;
        }
    }

    @Override
    public void update(Entity player, float dt) {
        checkActionPressed();
        relocate(dt);
        collisionObstacles(dt);
        Utils.getBrokerManager().detectionObservers.notifyDetection(velocity);
        player.send(new PositionEvent(currentPosition));
    }

    private void checkActionPressed() {
        if (isActionPressed) {
            Utils.getBrokerManager().actionObservers.notifyActionPressed(getCheckRect(), direction, currentPosition);
            isActionPressed = false;
        }
    }

    private void collisionObstacles(float dt) {
        if (velocity != Constant.MOVE_SPEED_4) {
            Utils.getBrokerManager().bumpObservers.notifyBump(getALittleBitBiggerBoundingBox(),
                                                              getCheckRect(), currentPosition);
            if (state.equals(EntityState.WALKING)) {
                collisionBlockers(dt);
                Utils.getBrokerManager().collisionObservers.notifyCollision(boundingBox, direction);
            }
        }
    }

    private void collisionBlockers(float dt) {
        List<Rectangle> blockers = Utils.getBrokerManager().blockObservers.getCurrentBlockersFor(boundingBox);
        if (!blockers.isEmpty()) {
            handleBlockers(blockers, dt);
        }
    }

    private void handleBlockers(List<Rectangle> blockers, float dt) {
        if (blockers.size() == 1) {
            moveSide(blockers.get(0), dt);
        }
        setRoundPosition();
        if (hasHitAnotherBlockWhileSideStepping()) {
            return;
        }
        while (doesBoundingBoxOverlapsBlockers()) {
            moveBack();
        }
    }

    private boolean hasHitAnotherBlockWhileSideStepping() {
        List<Rectangle> blockers = Utils.getBrokerManager().blockObservers.getCurrentBlockersFor(boundingBox);
        if (blockers.size() > 1) {
            currentPosition.set(oldPosition);
            setBoundingBox();
            return true;
        }
        return false;
    }

    private void moveSide(Rectangle blocker, float dt) {
        if (direction.equals(Direction.NORTH) || direction.equals(Direction.SOUTH)) {
            playerIsEastOf(blocker, dt);
            playerIsWestOf(blocker, dt);
        } else if (direction.equals(Direction.WEST) || direction.equals(Direction.EAST)) {
            playerIsSouthOf(blocker, dt);
            playerIsNorthOf(blocker, dt);
        }
    }

    private void playerIsEastOf(Rectangle blocker, float dt) {
        if (boundingBox.x + (boundingBox.width / 2) > blocker.x + blocker.width) {
            currentPosition.x += velocity * dt;
            if (boundingBox.x > blocker.x + blocker.width) {
                currentPosition.x = blocker.x + blocker.width;
            }
        }
    }

    private void playerIsWestOf(Rectangle blocker, float dt) {
        if (boundingBox.x + (boundingBox.width / 2) < blocker.x) {
            currentPosition.x -= velocity * dt;
            if (boundingBox.x + boundingBox.width < blocker.x) {
                currentPosition.x = blocker.x - boundingBox.width;
            }
        }
    }

    private void playerIsSouthOf(Rectangle blocker, float dt) {
        if (boundingBox.y + (boundingBox.height / 2) < blocker.y) {
            currentPosition.y -= velocity * dt;
            if (boundingBox.y + boundingBox.height < blocker.y) {
                currentPosition.y = blocker.y - boundingBox.height;
            }
        }
    }

    private void playerIsNorthOf(Rectangle blocker, float dt) {
        if (boundingBox.y + (boundingBox.height / 2) > blocker.y + blocker.height) {
            currentPosition.y += velocity * dt;
            if (boundingBox.y > blocker.y + blocker.height) {
                currentPosition.y = blocker.y + blocker.height;
            }
        }
    }

    private Rectangle getALittleBitBiggerBoundingBox() {
        Rectangle aLittleBitBiggerBox = new Rectangle(boundingBox);
        aLittleBitBiggerBox.setWidth(boundingBox.width + 24f);
        aLittleBitBiggerBox.setHeight(boundingBox.height + 24f);
        aLittleBitBiggerBox.setX(boundingBox.x - 12f);
        aLittleBitBiggerBox.setY(boundingBox.y - 12f);
        return aLittleBitBiggerBox;
    }

    private Rectangle getCheckRect() {
        Rectangle checkRect = new Rectangle();
        switch (direction) {
            case NORTH -> setNorth(checkRect);
            case SOUTH -> setSouth(checkRect);
            case WEST -> setWest(checkRect);
            case EAST -> setEast(checkRect);
            case NONE -> throw new IllegalArgumentException("Direction 'NONE' is not usable.");
        }
        return checkRect;
    }

    private void setNorth(Rectangle checkRect) {
        checkRect.setWidth(boundingBox.width / 4f);
        checkRect.setHeight(Constant.HALF_TILE_SIZE + 1);
        checkRect.setX(boundingBox.x + (boundingBox.width / 2f) - (boundingBox.width / 8f));
        checkRect.setY(boundingBox.y + boundingBox.height);
    }

    private void setSouth(Rectangle checkRect) {
        checkRect.setWidth(boundingBox.width / 4f);
        checkRect.setHeight(Constant.HALF_TILE_SIZE);
        checkRect.setX(boundingBox.x + (boundingBox.width / 2f) - (boundingBox.width / 8f));
        checkRect.setY(boundingBox.y - Constant.HALF_TILE_SIZE);
    }

    private void setWest(Rectangle checkRect) {
        checkRect.setWidth(Constant.HALF_TILE_SIZE);
        checkRect.setHeight(boundingBox.height + 6f);
        checkRect.setX(boundingBox.x - Constant.HALF_TILE_SIZE);
        checkRect.setY(boundingBox.y - 3f);
    }

    private void setEast(Rectangle checkRect) {
        checkRect.setWidth(Constant.HALF_TILE_SIZE);
        checkRect.setHeight(boundingBox.height + 6f);
        checkRect.setX(boundingBox.x + boundingBox.width);
        checkRect.setY(boundingBox.y - 3f);
    }

    @Override
    public void debug(ShapeRenderer shapeRenderer) {
        shapeRenderer.setColor(Color.YELLOW);
        shapeRenderer.rect(boundingBox.x, boundingBox.y, boundingBox.width, boundingBox.height);
        shapeRenderer.setColor(Color.GREEN);
        shapeRenderer.rect(getALittleBitBiggerBoundingBox().x,
                           getALittleBitBiggerBoundingBox().y,
                           getALittleBitBiggerBoundingBox().width,
                           getALittleBitBiggerBoundingBox().height);
        shapeRenderer.setColor(Color.CYAN);
        shapeRenderer.rect(getCheckRect().x,
                           getCheckRect().y,
                           getCheckRect().width,
                           getCheckRect().height);
    }

}

File: core/src/nl/t64/game/rpg/screens/world/entity/PhysicsPlayer.kt
package nl.t64.game.rpg.screens.world.entity

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.glutils.ShapeRenderer
import com.badlogic.gdx.math.Rectangle
import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.constants.Constant
import nl.t64.game.rpg.screens.world.entity.events.*


class PhysicsPlayer : PhysicsComponent() {

    private var isActionPressed = false

    init {
        boundingBoxWidthPercentage = 0.80f
        boundingBoxHeightPercentage = 0.30f
    }

    override fun receive(event: Event) {
        if (event is LoadEntityEvent) {
            direction = event.direction!!
            currentPosition = event.position
            setBoundingBox()
        }
        if (event is StateEvent) {
            state = event.state
        }
        if (event is DirectionEvent) {
            direction = event.direction
        }
        if (event is SpeedEvent) {
            velocity = event.moveSpeed
        }
        if (event is ActionEvent) {
            isActionPressed = true
        }
    }

    override fun update(player: Entity, dt: Float) {
        checkActionPressed()
        relocate(dt)
        collisionObstacles(dt)
        brokerManager.detectionObservers.notifyDetection(velocity)
        player.send(PositionEvent(currentPosition))
    }

    private fun checkActionPressed() {
        if (isActionPressed) {
            brokerManager.actionObservers.notifyActionPressed(getCheckRect(), direction, currentPosition)
            isActionPressed = false
        }
    }

    private fun collisionObstacles(dt: Float) {
        if (velocity != Constant.MOVE_SPEED_4) {
            brokerManager.bumpObservers.notifyBump(getALittleBitBiggerBoundingBox(), getCheckRect(), currentPosition)
            if (state == EntityState.WALKING) {
                collisionBlockers(dt)
            }
            brokerManager.collisionObservers.notifyCollision(boundingBox, direction)
        }
    }

    private fun collisionBlockers(dt: Float) {
        val blockers = brokerManager.blockObservers.getCurrentBlockersFor(boundingBox)
        if (blockers.isNotEmpty()) handleBlockers(blockers, dt)
    }

    private fun handleBlockers(blockers: List<Rectangle>, dt: Float) {
        if (blockers.size == 1) {
            moveSide(blockers[0], dt)
        }
        setRoundPosition()
        if (hasHitAnotherBlockWhileSideStepping()) {
            return
        }
        while (doesBoundingBoxOverlapsBlockers()) {
            moveBack()
        }
    }

    private fun hasHitAnotherBlockWhileSideStepping(): Boolean {
        val blockers = brokerManager.blockObservers.getCurrentBlockersFor(boundingBox)
        if (blockers.size > 1) {
            currentPosition.set(oldPosition)
            setBoundingBox()
            return true
        }
        return false
    }

    private fun moveSide(blocker: Rectangle, dt: Float) {
        if (direction == Direction.NORTH || direction == Direction.SOUTH) {
            playerIsEastOf(blocker, dt)
            playerIsWestOf(blocker, dt)
        } else if (direction == Direction.WEST || direction == Direction.EAST) {
            playerIsSouthOf(blocker, dt)
            playerIsNorthOf(blocker, dt)
        }
    }

    private fun playerIsEastOf(blocker: Rectangle, dt: Float) {
        if (boundingBox.x + (boundingBox.width / 2) > blocker.x + blocker.width) {
            currentPosition.x += velocity * dt
            if (boundingBox.x > blocker.x + blocker.width) {
                currentPosition.x = blocker.x + blocker.width
            }
        }
    }

    private fun playerIsWestOf(blocker: Rectangle, dt: Float) {
        if (boundingBox.x + (boundingBox.width / 2) < blocker.x) {
            currentPosition.x -= velocity * dt
            if (boundingBox.x + boundingBox.width < blocker.x) {
                currentPosition.x = blocker.x - boundingBox.width
            }
        }
    }

    private fun playerIsSouthOf(blocker: Rectangle, dt: Float) {
        if (boundingBox.y + (boundingBox.height / 2) < blocker.y) {
            currentPosition.y -= velocity * dt
            if (boundingBox.y + boundingBox.height < blocker.y) {
                currentPosition.y = blocker.y - boundingBox.height
            }
        }
    }

    private fun playerIsNorthOf(blocker: Rectangle, dt: Float) {
        if (boundingBox.y + (boundingBox.height / 2) > blocker.y + blocker.height) {
            currentPosition.y += velocity * dt
            if (boundingBox.y > blocker.y + blocker.height) {
                currentPosition.y = blocker.y + blocker.height
            }
        }
    }

    private fun getALittleBitBiggerBoundingBox(): Rectangle {
        return Rectangle(boundingBox).apply {
            setWidth(boundingBox.width + 24f)
            setHeight(boundingBox.height + 24f)
            setX(boundingBox.x - 12f)
            setY(boundingBox.y - 12f)
        }
    }

    private fun getCheckRect(): Rectangle {
        return when (direction) {
            Direction.NORTH -> getNorth()
            Direction.SOUTH -> getSouth()
            Direction.WEST -> getWest()
            Direction.EAST -> getEast()
            Direction.NONE -> throw IllegalArgumentException("Direction 'NONE' is not usable.")
        }
    }

    private fun getNorth(): Rectangle {
        return Rectangle().apply {
            setWidth(boundingBox.width / 4f)
            setHeight(Constant.HALF_TILE_SIZE + 1)
            setX(boundingBox.x + (boundingBox.width / 2f) - (boundingBox.width / 8f))
            setY(boundingBox.y + boundingBox.height)
        }
    }

    private fun getSouth(): Rectangle {
        return Rectangle().apply {
            setWidth(boundingBox.width / 4f)
            setHeight(Constant.HALF_TILE_SIZE)
            setX(boundingBox.x + (boundingBox.width / 2f) - (boundingBox.width / 8f))
            setY(boundingBox.y - Constant.HALF_TILE_SIZE)
        }
    }

    private fun getWest(): Rectangle {
        return Rectangle().apply {
            setWidth(Constant.HALF_TILE_SIZE)
            setHeight(boundingBox.height + 6f)
            setX(boundingBox.x - Constant.HALF_TILE_SIZE)
            setY(boundingBox.y - 3f)
        }
    }

    private fun getEast(): Rectangle {
        return Rectangle().apply {
            setWidth(Constant.HALF_TILE_SIZE)
            setHeight(boundingBox.height + 6f)
            setX(boundingBox.x + boundingBox.width)
            setY(boundingBox.y - 3f)
        }
    }

    override fun debug(shapeRenderer: ShapeRenderer) {
        shapeRenderer.color = Color.YELLOW
        shapeRenderer.rect(boundingBox.x, boundingBox.y, boundingBox.width, boundingBox.height)
        shapeRenderer.color = Color.GREEN
        shapeRenderer.rect(getALittleBitBiggerBoundingBox().x,
                           getALittleBitBiggerBoundingBox().y,
                           getALittleBitBiggerBoundingBox().width,
                           getALittleBitBiggerBoundingBox().height)
        shapeRenderer.color = Color.CYAN
        shapeRenderer.rect(getCheckRect().x,
                           getCheckRect().y,
                           getCheckRect().width,
                           getCheckRect().height)
    }

}
